testing/integration/tests/main.rs

```diff
@@ -1002,6 +1003,224 @@ fn test_oom4() {
+            ..Message::default()
+        };
+
+        let res = executor
+            .execute_message(message, ApplyKind::Explicit, 100)
+            .unwrap();
+        assert!(
+            res.msg_receipt.exit_code.is_success(),
+            "{:?}",
+            res.failure_info
+        );
+    }
+}
+
+#[test]
+fn readonly_actor_tests() {
+    // Instantiate tester
+    let mut tester = new_tester(
+        NetworkVersion::V21,
+        StateTreeVersion::V5,
+        MemoryBlockstore::default(),
+    )
+    .unwrap();
+
+    let [(_sender_id, sender_address)] = tester.create_accounts().unwrap();
+
+    let wasm_bin = READONLY_ACTOR_BINARY;
+
+    // Set actor state
+    let actor_state = [(); 0];
+    let state_cid = tester.set_state(&actor_state).unwrap();
+
+    // Set actor
+    let actor_address = Address::new_id(10000);
+
+    tester
+        .set_actor_from_bin(wasm_bin, state_cid, actor_address, TokenAmount::zero())
+        .unwrap();
+
+    // Instantiate machine
+    tester.instantiate_machine(DummyExterns).unwrap();
+
+    let executor = tester.executor.as_mut().unwrap();
+
+    let message = Message {
+        from: sender_address,
+        to: actor_address,
+        gas_limit: 1000000000,
+        method_num: 2,
+        sequence: 0,
+        value: TokenAmount::from_atto(100),
+        ..Message::default()
+    };
+
+    let res = executor
+        .execute_message(message, ApplyKind::Explicit, 100)
+        .unwrap();
+    assert!(
+        res.msg_receipt.exit_code.is_success(),
+        "{:?}",
+        res.failure_info
+    );
+    assert!(res.msg_receipt.events_root.is_none());
+}
+
+#[test]
+fn upgrade_actor_test() {
+    let mut tester = new_tester(
+        NetworkVersion::V21,
+        StateTreeVersion::V5,
+        MemoryBlockstore::default(),
+    )
+    .unwrap();
+
+    let sender: [Account; 4] = tester.create_accounts().unwrap();
+    let receiver = Address::new_id(10000);
+    let state_cid = tester.set_state(&[(); 0]).unwrap();
+
+    let wasm_bin = UPGRADE_ACTOR_BINARY;
+    tester
+        .set_actor_from_bin(wasm_bin, state_cid, receiver, TokenAmount::zero())
+        .unwrap();
+    tester.instantiate_machine(DummyExterns).unwrap();
+
+    let executor = tester.executor.as_mut().unwrap();
+
+    {
+        // test a successful call to `upgrade` endpoint
+        let message = Message {
+            from: sender[0].1,
+            to: receiver,
+            gas_limit: 1000000000,
+            method_num: 1,
+            sequence: 0_u64,
+            value: TokenAmount::from_atto(100),
+            ..Message::default()
+        };
+
+        let res = executor
+            .execute_message(message, ApplyKind::Explicit, 100)
+            .unwrap();
+        assert!(
+            res.msg_receipt.exit_code.is_success(),
+            "{:?}",
+            res.failure_info
+        );
+        let val: i64 = res.msg_receipt.return_data.deserialize().unwrap();
+        assert_eq!(val, 666);
+    }
+
+    {
```

Comment:
Let's document what each of these cases is doing.

Reply:
Note that this was already documented in the upgrade-actor